spynnaker/pyNN/models/neuron/population_vertex.py

```diff
@@ -474,35 +498,195 @@ def set_structure(self, structure: BaseStructure) -> None:
+                    n_conns = dynamics.pad_to_length
+                if n_conns is None:
+                    n_conns = conn.get_n_connections_from_pre_vertex_maximum(
+                        self.get_max_atoms_per_core(), s_info)
+                # The number of synapses is the number of connections from each
+                # pre-neuron to each post-neuron
+                assert n_conns is not None
+                pre_synapses_per_second += math.ceil(
+                    n_conns * spikes_per_second * pre_vertex.n_atoms)
+
+            if self._is_direct_poisson(pre_vertex, projs):
+                poisson_synapses_per_second += pre_synapses_per_second
+            else:
+                synapses_per_second += pre_synapses_per_second
+
+        # How many cores are needed to process the non-direct-poisson synapses?
+        n_synapse_cores = math.ceil(
+            synapses_per_second / synapses_per_core_per_sim_second)
+
+        # The number of Poisson cores that will be needed
+        n_poisson_cores = len(self.incoming_poisson_projections)
+
+        # If we can definitely do the Poisson vertices directly,
+        # lets recommend this
+        if n_synapse_cores + n_poisson_cores < max_n_cores:
+            self.__n_synapse_cores = n_synapse_cores
+        else:
+
+            # Otherwise, we should consider how many more cores we need for the
+            # Poisson input spikes
+            n_synapse_cores = math.ceil(
+                (synapses_per_second + poisson_synapses_per_second) /
+                synapses_per_core_per_sim_second)
+
+            # If the number of cores needed is more than the maximum, use the
+            # maximum
+            self.__n_synapse_cores = min(max_n_cores, n_synapse_cores)
+
+    @property
+    def max_delay_steps(self) -> int:
+        """
+        The maximum number of delay steps supported on a core.
+        """
+        self.__update_max_delay()
+        assert self.__max_delay_slots_available is not None
+        return self.__max_delay_slots_available
+
+    @property
+    def max_delay_steps_incoming(self) -> int:
+        """
+        The maximum delay steps needed to handle incoming synapses,
+        accounting for delay extensions.
+        """
+        self.__update_max_delay()
+        assert self.__max_delay_ms is not None
+        assert self.__max_delay_slots_available is not None
+        max_incoming_slots = 2 ** get_n_bits(math.ceil(
+            self.__max_delay_ms /
+            SpynnakerDataView().get_simulation_time_step_ms()))
+        return min(max_incoming_slots, self.__max_delay_slots_available)
+
+    @property
+    def allow_delay_extension(self) -> bool:
+        """
+        Whether delay extension should be allowed or not.
+        """
+        if not self.__allow_delay_extensions:
+            return False
+        # Determine if we *expect* a delay extension; if not disallow
+        self.__update_max_delay()
```

Comment:
use self.def max_delay_steps

Reply:
Could do for one variable, but need both; each of these methods are subtly different e.g. max_delay_steps is how many steps are available for use given other things, max_delay_steps_incoming are how many we actually need for the incoming synapses, with delay extensions accounted for, and allow_delay_extensions is checking if the maximum delay of the incoming synapses without extensions is more than the maximum number available for use.

Reply:
Now updated to use the return values